Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean

```diff
@@ -0,0 +1,682 @@
+that `max I₁ < ... < max Iₖ`, and let `iₘ` be the number of elements of `Iₘ`. Then
+`D^n (g ∘ f) (x) (v_1, ..., v_n) =
+  ∑_{I partition of {1, ..., n}} D^k g (f x) (D^{i₁} f (x) (v_{I₁}), ..., D^{iₖ} f (x) (v_{Iₖ}))`
+where by `v_{Iₘ}` we mean the vectors `vᵢ` with indices in `Iₘ`.
+
+For instance, for `n = 2`, there are 2 partitions of `{1, 2}`, given by `{1} {2}` and `{1, 2}`,
+and therefore
+`D^2(g ∘ f) (x) (v₁, v₂) = D^2 g (f x) (Df (x) v₁, Df (x) v₂) + Dg (f x) (D^2f (x) (v₁, v₂))`.
+
+The formula is straightforward to prove by induction, as differentiating
+`D^k g (f x) (D^{i₁} f (x) (v_{I₁}), ..., D^{i_k} f (x) (v_{Iₖ}))` gives a sum with `k + 1` terms
+where one differentiates either `D^k g (f x)`, or one of the `D^{iₘ} f (x)`, amounting to
+adding to the partition `I` either a new atom `{0}` to its left, or extending `Iₘ` by adding `0`
+to it. In this way, one obtains bijectively all partitions of `{0, ..., n}`, and the proof can
+go on (up to relabelling).
+
+The main difficulty is to write down things in a precise language, namely to write
+`D^k g (f x) (D^{i₁} f (x) (v_{I₁}), ..., D^{iₖ} f (x) (v_{Iₖ}))` as a continuous multilinear map
+of the `vᵢ`. For this, instead of working with partitions of `{1, ..., n}` and ordering their
+parts, we work with partitions in which the ordering is part of the data -- this is equivalent,
+but much more convenient to implement. We call these `OrderedFinpartition n`.
+
+## Main results
+
+Given `c : OrderedFinpartition n` and two formal multilinear series `q` and `p`, we
+define `c.compAlongOrderedFinpartition q p` as an `n`-multilinear map given by the formula above,
+i.e., `(v₁, ..., vₙ) ↦ qₖ (p_{i₁} (v_{I₁}), ..., p_{iₖ} (v_{Iₖ}))`.
+
+Then, we define `q.taylorComp p` as a formal multilinear series whose `n`-th term is
+the sum of `c.compAlongOrderedFinpartition q p` over all ordered finpartitions of size `n`.
+
+Finally, we prove in `HasFTaylorSeriesUptoOn.comp` that, if two functions `g` and `f` have Taylor
+series up to `n` given by `q` and `p`, then `g ∘ f` also has a Taylor series,
+given by `q.taylorComp p`.
+
+## Implementation
+
+A first technical difficulty is to implement the extension process of `OrderedFinpartition`
+corresponding to adding a new atom, or appending an atom to an existing part, and defining the
+associated increasing parameterizations that show up in the definition
+of `compAlongOrderedFinpartition`.
+
+Then, one has to show that the ordered finpartitions thus
+obtained give exactly all ordered finpartitions of order `n+1`. For this, we define the inverse
+process (shrinking a finpartition of `n+1` by erasing `0`, either as an atom or from the part
+that contains it), and we show that these processes are inverse to each other, yielding an
+equivalence between `((c : OrderedFinpartition n) × Option (Fin c.length))`
+and `OrderedFinpartition (n + 1)`. This equivalence shows up prominently in the inductive proof
+of Faa di Bruno formula to identify the sums that show up.
```

Comment:
Yes, it's definitely an unusual implementation. The reason is really that it's the one I need for the Faa di Bruno formula. The formula reads as follows (mixing math and Lean, but hopefully this should be clear) :
```
D^n (g o f) (x) (v_0, ..., v_{n-1}) = \sum_{c : OrderedFinpartition n}, D^c.length g (f x, fun (i : Fin c.length) -> D^(c.partSize i) f (v o c.emb i))
```
So, you can see that the `OrderedFinpartition` shows up through the embeddings `c.emb`. The formula is only true if the parts are ordered by increasing maxima (in the general situation where the iterated derivatives are not symmetric a priori). And in the inductive proof of the formula, what matters is how the embeddings correspond to each other when one extend the finpartition by adding an atom to the left, hence the implementation of `extendEquiv` which builds this with nice defeqs.

I don't expect this very specific implementation can be used for anything else than the Faa di Bruno formula, which is why I have left it in this file instead of creating a new file in the `Combinatorics` folder. You can have a look at https://github.com/leanprover-community/mathlib4/pull/17152/files to see how the completed file looks like, and how the implementation is used in the inductive proof of the formula.

Reply:
I see, that makes sense. Could you add your second paragraph - or something similar to it - to the implementation notes in this file? Similarly it might be nice to add to the docstring of OrderedFinpartition saying that it is only intended to be used for this formula, to prevent future contributors being confused.

My only remaining worry is that the implementation details of OrderedFinpartition may need to be exposed to people using the Faa di Bruno formula, since as you say, the formula genuinely uses the embeddings it contains, but it looks like your `taylorComp` abstraction hides this nicely?

Reply:
I have expanded the docstrings (both file-level and of the structure) to explain the design, and how specific it is.

For my application, it is indeed not necessary to look at the details of ordered finpartitions, the API is hiding this nicely. For other applications, it could be useful to have a bound on the cardinality of ordered finpartitions, and for that constructing the equiv with usual Finpartitions will be useful. But I'd say this can wait until someone needs it.